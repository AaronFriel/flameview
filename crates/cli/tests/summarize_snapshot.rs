#![cfg(not(miri))]

use std::path::{Path, PathBuf};
use std::process::Command;

use regex::Regex;

fn run_summary(fixture: &Path, max_lines: usize, coverage: f64) -> String {
    let bin = env!("CARGO_BIN_EXE_flameview");
    let output = Command::new(bin)
        .arg(fixture)
        .arg("--summarize")
        .arg("--max-lines")
        .arg(max_lines.to_string())
        .arg("--coverage")
        .arg(coverage.to_string())
        .output()
        .expect("run");
    assert!(
        output.status.success(),
        "{}",
        String::from_utf8_lossy(&output.stderr)
    );
    let stdout = String::from_utf8(output.stdout).expect("utf8");
    let re = Regex::new(r"[A-Z]:[/\\][^\s]*").expect("regex");
    let cleaned = re.replace_all(&stdout, "");
    let cleaned = cleaned.replace("\r\n", "\n");
    let lines: Vec<_> = cleaned.lines().map(|l| l.trim_end()).collect();
    lines.join("\n").trim_end_matches('\n').to_string()
}

#[test]
fn summarize_snapshots() {
    let data_dir: PathBuf = [env!("CARGO_MANIFEST_DIR"), "..", "..", "tests", "data"]
        .into_iter()
        .collect();

    let cases = [
        ("small_default", data_dir.join("small.txt"), 50, 0.95),
        ("medium_30_90", data_dir.join("medium.txt"), 30, 0.90),
        ("large_15_80", data_dir.join("large.txt"), 15, 0.80),
    ];

    for (name, fixture, lines, cov) in cases {
        let output = run_summary(&fixture, lines, cov);
        let mut settings = insta::Settings::new();
        settings.set_prepend_module_to_snapshot(false);
        settings.bind(|| insta::assert_snapshot!(name, output));
    }
}

#[test]
#[should_panic]
fn summarize_corrupt_panics() {
    let bad: PathBuf = [env!("CARGO_MANIFEST_DIR"), "..", "..", "README.md"]
        .into_iter()
        .collect();

    run_summary(&bad, 10, 0.5);
}
