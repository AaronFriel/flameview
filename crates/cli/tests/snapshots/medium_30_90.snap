---
source: crates/cli/tests/summarize_snapshot.rs
expression: output
---
(root) (100.0%, 223)
  java (100.0%, 223)
    start_thread (99.6%, 222)
      java_start (99.6%, 222)
        JavaThread::run (93.3%, 208)
          JavaThread::thread_main_inner (93.3%, 208)
            thread_entry (93.3%, 208)
              JavaCalls::call_virtual (93.3%, 208)
                JavaCalls::call_virtual (93.3%, 208)
                  JavaCalls::call_helper (93.3%, 208)
                    call_stub_[j] (93.3%, 208)
                      Interpreter_[j] (93.3%, 208)
                        Interpreter_[j] (93.3%, 208)
                          io/netty/channel/nio/NioEventLoop:.run_[j] (93.3%, 208)
                            io/netty/channel/nio/NioEventLoop:.processSelectedKeys_[j] (92.8%, 207)
                              io/netty/channel/nio/NioEventLoop:.processSelectedKeysOptimized_[j] (92.8%, 207)
                                io/netty/channel/nio/NioEventLoop:.processSelectedKey_[j] (92.4%, 206)
                                  io/netty/channel/nio/AbstractNioByteChannel$NioByteUnsafe:.read_[j] (92.4%, 206)
                                    io/netty/channel/AbstractChannelHandlerContext:.fireChannelRead_[j] (47.5%, 106)
                                      io/netty/handler/codec/ByteToMessageDecoder:.channelRead_[j] (47.5%, 106)
                                        io/netty/channel/AbstractChannelHandlerContext:.fireChannelRead_[j] (46.6%, 104)
                                          org/vertx/java/core/net/impl/VertxHandler:.channelRead_[j] (46.6%, 104)
                                            org/vertx/java/core/http/impl/DefaultHttpServer$ServerHandler:.doMessageReceived_[j] (44.8%, 100)
                                              org/mozilla/javascript/gen/file__home_bgregg_testtest_vert_x_2_1_4_sys_mods_io_vertx_lang_js_1_1_0_[j] (41.7%, 93)
                                                org/mozilla/javascript/gen/file__home_bgregg_testtest_vert_x_2_1_4_sys_mods_io_vertx_lang_js_1_1_0_[j] (41.3%, 92)
                                                  org/mozilla/javascript/gen/file__home_bgregg_testtest_vert_x_2_1_4_sys_mods_io_vertx_lang_js_1_1_0_[j] (36.8%, 82)
                                                    org/mozilla/javascript/gen/file__home_bgregg_testtest_vert_x_2_1_4_sys_mods_io_vertx_lang_js_1_1_0_[j] (36.3%, 81)
                                                      org/mozilla/javascript/ScriptRuntime:.newObject_[j] (14.3%, 32)
                                                        org/mozilla/javascript/gen/file__home_bgregg_testtest_vert_x_2_1_4_sys_mods_io_vertx_lang_js_1_1_0_[j] (14.3%, 32)
                                                          org/mozilla/javascript/gen/file__home_bgregg_testtest_vert_x_2_1_4_sys_mods_io_vertx_lang_js_1_1_0_[j] (14.3%, 32)
                                                            org/mozilla/javascript/optimizer/OptRuntime:.call2_[j] (5.8%, 13)
